File: NotesApp/src/components/NoteItem.js
import React, { useState } from "react";
import { View, Text, StyleSheet, TouchableOpacity, Alert } from "react-native";
import { deleteNote } from "../services/note.service";

const NoteItem = ({ note, onNoteDeleted, onEdit }) => {
  const [deleting, setDeleting] = useState(false);

  const formatDate = (dateString) => {
    const date = new Date(dateString);
    return date.toLocaleDateString();
  };

  const handleDelete = () => {
    Alert.alert("Delete Note", "Are you sure you want to delete this note?", [
      { text: "Cancel", style: "cancel" },
      {
        text: "Delete",
        style: "destructive",
        onPress: async () => {
          try {
            setDeleting(true);
            await deleteNote(note.$id);
            if (onNoteDeleted) onNoteDeleted(note.$id);
          } catch (error) {
            Alert.alert("Error", "Failed to delete note. Please try again.");
          } finally {
            setDeleting(false);
          }
        },
      },
    ]);
  };

  return (
    <View style={styles.container}>
      <View style={styles.content}>
        <Text style={styles.title}>{note.title}</Text>
        <Text style={styles.date}>Last updated: {formatDate(note.updatedAt)}</Text>
        <Text style={styles.noteContent} numberOfLines={3}>{note.content}</Text>
      </View>
      <View style={styles.buttonContainer}>
        <TouchableOpacity style={styles.editButton} onPress={() => onEdit(note)}>
          <Text style={styles.editText}>Edit</Text>
        </TouchableOpacity>
        <TouchableOpacity style={styles.deleteButton} onPress={handleDelete} disabled={deleting}>
          <Text style={styles.deleteText}>{deleting ? "..." : "Delete"}</Text>
        </TouchableOpacity>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: { backgroundColor: "white", borderRadius: 8, padding: 16, marginHorizontal: 16, marginBottom: 12, shadowColor: "#000", shadowOffset: { width: 0, height: 1 }, shadowOpacity: 0.2, shadowRadius: 2, elevation: 2 },
  content: { flex: 1 },
  title: { fontSize: 18, fontWeight: "bold", marginBottom: 4 },
  date: { fontSize: 12, color: "#666", marginBottom: 8 },
  noteContent: { fontSize: 14, color: "#333" },
  buttonContainer: { flexDirection: "row", justifyContent: "flex-end", marginTop: 12, borderTopWidth: 1, borderTopColor: "#eee", paddingTop: 8 },
  editButton: { marginRight: 16 },
  editText: { color: "#2196F3", fontWeight: "500" },
  deleteButton: {},
  deleteText: { color: "red", fontWeight: "500" },
});

export default NoteItem;

File: notes_app/lib/widgets/note_item.dart
import 'package:flutter/material.dart';
import 'package:appwrite/models.dart';
import '../services/note_service.dart';

class NoteItem extends StatefulWidget {
  final Document note;
  final Function(String)? onNoteDeleted;
  final VoidCallback? onEdit;

  const NoteItem({Key? key, required this.note, this.onNoteDeleted, this.onEdit}) : super(key: key);

  @override
  _NoteItemState createState() => _NoteItemState();
}

class _NoteItemState extends State<NoteItem> {
  final NoteService _noteService = NoteService();
  bool _isDeleting = false;

  String _formatDate(String dateString) {
    final date = DateTime.parse(dateString);
    return '${date.day}/${date.month}/${date.year}';
  }

  Future<void> _handleDelete() async {
    final confirmed = await showDialog<bool>(
      context: context,
      builder: (ctx) => AlertDialog(
        title: const Text('Delete Note'),
        content: const Text('Are you sure you want to delete this note?'),
        actions: [
          TextButton(onPressed: () => Navigator.of(ctx).pop(false), child: const Text('Cancel')),
          TextButton(onPressed: () => Navigator.of(ctx).pop(true), child: const Text('Delete', style: TextStyle(color: Colors.red))),
        ],
      ),
    );

    if (confirmed == true) {
      try {
        setState(() { _isDeleting = true; });
        await _noteService.deleteNote(widget.note.$id);
        if (widget.onNoteDeleted != null) widget.onNoteDeleted!(widget.note.$id);
      } catch (e) {
        if (mounted) {
          ScaffoldMessenger.of(context).showSnackBar(
            const SnackBar(content: Text('Failed to delete note.'), backgroundColor: Colors.red),
          );
        }
      } finally {
        if (mounted) setState(() { _isDeleting = false; });
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    final title = widget.note.data['title'] as String? ?? '';
    final content = widget.note.data['content'] as String? ?? '';
    final updatedAt = widget.note.data['updatedAt'] as String? ?? widget.note.$updatedAt;

    return Card(
      elevation: 2,
      margin: const EdgeInsets.only(bottom: 12),
      child: Padding(
        padding: const EdgeInsets.all(16.0),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Text(title, style: const TextStyle(fontSize: 18, fontWeight: FontWeight.bold)),
            const SizedBox(height: 4),
            Text('Last updated: ${_formatDate(updatedAt)}', style: const TextStyle(fontSize: 12, color: Colors.grey)),
            const SizedBox(height: 8),
            Text(content, style: const TextStyle(fontSize: 14), maxLines: 3, overflow: TextOverflow.ellipsis),
            const Divider(height: 20),
            Row(
              mainAxisAlignment: MainAxisAlignment.end,
              children: [
                TextButton(onPressed: widget.onEdit, child: const Text('Edit')),
                const SizedBox(width: 8),
                TextButton(
                  onPressed: _isDeleting ? null : _handleDelete,
                  child: _isDeleting
                      ? const SizedBox(width: 16, height: 16, child: CircularProgressIndicator(strokeWidth: 2))
                      : const Text('Delete', style: TextStyle(color: Colors.red)),
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }
}
